### ISSUE TITLE
Web Sockets Streamed over Https, multiple errors

### ISSUE THREAD

**Issue by StephenBonikowsky** [dotnet/wcf#470]
The test case...
WebSocketHttpsDuplexBinaryStreamed located...
C:\OSS\wcf\src\System.Private.ServiceModel\tests\Scenarios\Extensibility\WebSockets\WebSocketTests.cs

Is failing with two issues, the first one throws a "System.PlatformNotSupportedException : Channel shape not supported" from...
C:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\HttpsChannelFactory.cs
Line 78.
Replace with...
return (TChannel)(object)new ClientWebSocketTransportDuplexSessionChannel((HttpChannelFactory<IDuplexSessionChannel>)(object)this, this.ClientWebSocketFactory, address, via, this.WebSocketBufferPool);

Next exception is...
System.ArgumentException occurred
  Message=Empty string is not a valid subprotocol value. Please use \"null\" to specify no value.
Parameter name: subProtocol
  ParamName=subProtocol
  Source=System.Net.WebSockets.Client
  StackTrace:
       at System.Net.WebSockets.WebSocketValidate.ValidateSubprotocol(String subProtocol)

From...
C:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\CoreClrClientWebSocketFactory.cs
Line 19

**Comment by roncain** [comment]
Looks like PR https://github.com/dotnet/wcf/pull/472 is addressing the HttpChannelFactory part of this.  Thanks @StephenBonikowsky .

**Comment by mconnew** [comment]
The current problem is ClientWebSocket is throwing an exception on the connect call with an error saying it was unable to connect to the remote host.

**Comment by roncain** [comment]
Thanks @mconnew.  What is the next step?  Are you preparing a PR with a fix?

**Comment by mconnew** [comment]
Secure websockets streaming tests are failing because of a bug in WebSockets. See dotnet/corefx#4429

**Comment by mconnew** [comment]
These tests are passing.

**Comment by roncain** [comment]
Reopening this issue because this test does not pass for me when I run locally self-hosted.   I set PSEXEC_PATH to the folder for psexec.exe, set IncludeTestsWithIssues=true, and run all OuterLoops. Note that Issue(470) is still applied to this test so it is not being run by CI or Helix.  Stack trace is:
```
            WebSocketTests.WebSocket_Https_Duplex_TextStreamed [FAIL]
               System.TimeoutException : This request operation sent to wss://localhost:8084//WebSocketHttpsDuplexTextStreamed.svc//WebSocketHttpsDuplexTextStreamedResource did not receive a reply within the configured timeout (00:00:59.9990033).  The time allotted to this operation may have been a portion of a longer timeout.  This may be because the service is still processing the operation or because the service was unable to send a reply message.  Please consider increasing the operation timeout (by casting the channel/proxy to IContextChannel and setting the OperationTimeout property) and ensure that the service is able to connect to the client.
         ---ETW Trace for Test WebSocketTests.WebSocket_Https_Duplex_TextStreamed Begins---
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: ChannelFactory created .
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: ServiceChannelCall started.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Client request serialization started.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Client completed serialization of the request message.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: ServiceChannel Open Start.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocket connection request to 'wss://localhost:8084//WebSocketHttpsDuplexTextStreamed.svc//WebSocketHttpsDuplexTextStreamedResource' send start.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Creating the client WebSocket with a factory of type 'System.ServiceModel.Channels.CoreClrClientWebSocketFactory'.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Pool allocating 16384 Bytes.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 read start.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 connection request sent.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: ChannelHandlerId:19385132 - Message receive loop started.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: ServiceChannel Open Stop.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: TextMessageEncoder started encoding the message.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 writing '402' bytes to ''.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 asynchronous write stop.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 writing '193' bytes to ''.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 asynchronous write stop.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: A streamed message was written by the encoder.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 writing '0' bytes to ''.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 asynchronous write stop.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: The transport sent a message to 'wss://localhost:8084//WebSocketHttpsDuplexTextStreamed.svc//WebSocketHttpsDuplexTextStreamedResource'.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: TextMessageEncoder started decoding the message.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 read start.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Throwing an exception. Source: System.ServiceModel 4.0.0.0. Exception details: System.TimeoutException: This request operation sent to wss://localhost:8084//WebSocketHttpsDuplexTextStreamed.svc//WebSocketHttpsDuplexTextStreamedResource did not receive a reply within the configured timeout (00:00:59.9990033).  The time allotted to this operation may have been a portion of a longer timeout.  This may be because the service is still processing the operation or because the service was unable to send a reply message.  Please consider increasing the operation timeout (by casting the channel/proxy to IContextChannel and setting the OperationTimeout property) and ensure that the service is able to connect to the client.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 sending close output message to 'NormalClosure' with close status 'wss://localhost:8084//WebSocketHttpsDuplexTextStreamed.svc//WebSocketHttpsDuplexTextStreamedResource'.
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: Throwing an exception. Source: System.ServiceModel 4.0.0.0. Exception details: System.TimeoutException: Receive timed out after 00:00:59.9989965. The time allotted to this operation may have been a portion of a longer timeout. ---> System.TimeoutException: Cannot claim lock within the allotted timeout of 00:00:59.9989965. The time allotted to this operation may have been a portion of a longer timeout.
            --- End of inner exception stack trace ---
         WebSocketTests.WebSocket_Https_Duplex_TextStreamed: WebSocketId:48858004 connection is aborted.
         ---ETW Trace for Test WebSocketTests.WebSocket_Https_Duplex_TextStreamed Ends---
         
               Stack Trace:
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Dispatcher\DuplexChannelBinder.cs(771,0): at System.ServiceModel.Dispatcher.DuplexChannelBinder.SyncDuplexRequest.WaitForReply(TimeSpan timeout)
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Dispatcher\DuplexChannelBinder.cs(524,0): at System.ServiceModel.Dispatcher.DuplexChannelBinder.Request(Message message, TimeSpan timeout)
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\ServiceChannel.cs(765,0): at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\ServiceChannel.cs(722,0): at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs)
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\ServiceChannelProxy.cs(369,0): at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(MethodCall methodCall, ProxyOperationRuntime operation)
                  c:\OSS\wcf\src\System.Private.ServiceModel\src\System\ServiceModel\Channels\ServiceChannelProxy.cs(141,0): at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(MethodInfo targetMethod, Object[] args)
                  --- End of stack trace from previous location where exception was thrown ---
                     at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                     at System.Reflection.DispatchProxyGenerator.Invoke(Object[] args)
                     at generatedProxy_1.DownloadStream()
                  c:\OSS\wcf\src\System.Private.ServiceModel\tests\Scenarios\Extensibility\WebSockets\WebSocketTests.4.1.0.cs(330,0): at WebSocketTests.WebSocket_Https_Duplex_TextStreamed()
         Finished:    Extensibility.WebSockets.Tests
```

**Comment by mconnew** [comment]
I have confirmed this is a reliable reproducible instance of dotnet/corefx#4429

**Comment by roncain** [comment]
Thanks @mconnew -- it looks like https://github.com/dotnet/corefx/issues/4429 is slotted to be fixed in 1.2

**Comment by zhenlan** [comment]
@mconnew please check if the corefx fix is included in 1.1 branch.

**Comment by zhenlan** [comment]
@mconnew has ported the fix made in corefx already make in 1.1. The remaining task is to re-enable the tests that are disabled by this issue.

**Comment by StephenBonikowsky** [comment]
Seems to be working now. Closing.
